Exclude unsupported types from automatic timestamp and TTL preservation

Currently when unloading using the options `-timestamp` or `-ttl` (to automatically preserve cell timestamp and TTL) the operation will fail if the table being unloaded contains collections.

With OSS Cassandra, the operation fails right away because the generated query is invalid. The selector functions `writetime()` and `ttl()` cannot be applied to collections (even frozen ones, see [CASSANDRA-14533](https://issues.apache.org/jira/browse/CASSANDRA-14533)):

```
2022-02-16 14:27:36 ERROR Operation UNLOAD_TIMESTAMP_TTL failed: Cannot use selection function writeTime on collections.
com.datastax.oss.driver.api.core.servererrors.InvalidQueryException: Cannot use selection function writeTime on collections
	at com.datastax.oss.driver.api.core.servererrors.InvalidQueryException.copy(InvalidQueryException.java:48)
	at com.datastax.oss.driver.internal.core.util.concurrent.CompletableFutures.getUninterruptibly(CompletableFutures.java:149)
	at com.datastax.oss.driver.internal.core.cql.CqlPrepareSyncProcessor.process(CqlPrepareSyncProcessor.java:59)
	at com.datastax.oss.driver.internal.core.cql.CqlPrepareSyncProcessor.process(CqlPrepareSyncProcessor.java:31)
	at com.datastax.oss.driver.internal.core.session.DefaultSession.execute(DefaultSession.java:230)
	at com.datastax.oss.driver.api.core.cql.SyncCqlSession.prepare(SyncCqlSession.java:224)
	at com.datastax.oss.dsbulk.workflow.commons.settings.SchemaSettings.prepareStatementAndCreateMapping(SchemaSettings.java:720)
	at com.datastax.oss.dsbulk.workflow.commons.settings.SchemaSettings.createReadResultMapper(SchemaSettings.java:510)
	at com.datastax.oss.dsbulk.workflow.unload.UnloadWorkflow.init(UnloadWorkflow.java:150)
	at com.datastax.oss.dsbulk.runner.WorkflowThread.run(WorkflowThread.java:52)
```

With DSE, these functions are allowed to be applied to collections; but they return a list of BIGINTs (one element per collection item). The operation then fails during the row mapping phase with:

```
java.lang.IllegalArgumentException: Could not deserialize column "writetime(contact)" of type List(BIGINT, not frozen) as java.lang.String
	at com.datastax.oss.dsbulk.workflow.commons.schema.DefaultReadResultMapper.map(DefaultReadResultMapper.java:77)
	at java.lang.Thread.run(Thread.java:748) [17 skipped]
Caused by: java.lang.IllegalArgumentException: Cannot create a WriteTimeCodec for List(BIGINT, not frozen)
	at com.datastax.oss.dsbulk.mapping.DefaultMapping.createWritetimeCodec(DefaultMapping.java:101)
	at com.datastax.oss.dsbulk.mapping.DefaultMapping.lambda$codec$0(DefaultMapping.java:90)
```

We could improve this:

* If the server does not support unloading timestamp and TTL of collection columns (OSS C*):
  * Simply do not export it
  * Log a warning that timestamp and TTL data will not be preserved for these columns, and keep going.
* If the server supports it (DSE), then we should be able to export at least one value of the list. It's not possible to specify individual element timestamps and TTLs when loading anyways, so exporting one single value is the best we can do.
  * Modify `WriteTimeCodec` to also accept `list<bigint>`, and only retain the first element. 
  * For TTL variables we'll need to intercept them the same way we intercept writetime variables. Then, only retain the first element.
  * Log a warning explaining that some timestamps and TTLs may be lost.

┆Issue is synchronized with this [Jira Task](https://datastax-oss.atlassian.net/browse/BULK-25) by [Unito](https://www.unito.io)

--- a/changelog/README.md
+++ b/changelog/README.md
@@ -5,6 +5,7 @@
 - [improvement] Upgrade driver to 4.14.0.
 - [new feature] [#400](https://github.com/datastax/dsbulk/issues/400): Add ability to unwrap BATCH queries.
 - [new feature] [#405](https://github.com/datastax/dsbulk/issues/405): Add support for Prometheus.
+- [improvement] [#403](https://github.com/datastax/dsbulk/issues/403): Exclude unsupported types from automatic timestamp and TTL preservation.
 
 ## 1.8.0
 

--- a/runner/src/it/java/com/datastax/oss/dsbulk/runner/ccm/CSVConnectorEndToEndCCMIT.java
+++ b/runner/src/it/java/com/datastax/oss/dsbulk/runner/ccm/CSVConnectorEndToEndCCMIT.java
@@ -49,11 +49,20 @@ import com.datastax.oss.driver.api.core.DefaultProtocolVersion;
 import com.datastax.oss.driver.api.core.Version;
 import com.datastax.oss.driver.api.core.cql.ResultSet;
 import com.datastax.oss.driver.api.core.cql.Row;
+import com.datastax.oss.driver.api.core.data.CqlDuration;
 import com.datastax.oss.driver.api.core.data.TupleValue;
 import com.datastax.oss.driver.api.core.data.UdtValue;
+import com.datastax.oss.driver.api.core.type.DataType;
+import com.datastax.oss.driver.api.core.type.DataTypes;
+import com.datastax.oss.driver.api.core.type.TupleType;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
+import com.datastax.oss.driver.api.core.type.codec.TypeCodec;
+import com.datastax.oss.driver.api.core.type.codec.registry.CodecRegistry;
+import com.datastax.oss.driver.api.core.uuid.Uuids;
 import com.datastax.oss.driver.shaded.guava.common.base.Splitter;
 import com.datastax.oss.driver.shaded.guava.common.base.Strings;
 import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableSet;
 import com.datastax.oss.driver.shaded.guava.common.collect.Lists;
 import com.datastax.oss.dsbulk.codecs.api.util.CodecUtils;
 import com.datastax.oss.dsbulk.codecs.api.util.OverflowStrategy;
@@ -74,13 +83,18 @@ import com.datastax.oss.dsbulk.tests.logging.StreamInterceptor;
 import com.datastax.oss.dsbulk.tests.utils.CQLUtils;
 import com.datastax.oss.dsbulk.tests.utils.FileUtils;
 import com.datastax.oss.dsbulk.workflow.api.log.OperationDirectory;
+import com.google.common.collect.ImmutableList;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.LineNumberReader;
 import java.math.BigDecimal;
+import java.math.BigInteger;
 import java.math.RoundingMode;
+import java.net.InetAddress;
 import java.net.URI;
 import java.net.URL;
+import java.net.UnknownHostException;
+import java.nio.ByteBuffer;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.time.Instant;
@@ -98,12 +112,15 @@ import java.util.stream.Stream;
 import org.assertj.core.api.InstanceOfAssertFactories;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Assumptions;
 import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Tag;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.CsvSource;
+import org.junit.jupiter.params.provider.MethodSource;
 
 @CCMConfig(numberOfNodes = 1, config = "enable_user_defined_functions:true")
 @Tag("medium")
@@ -1953,21 +1970,30 @@ class CSVConnectorEndToEndCCMIT extends EndToEndCCMITBase {
     assertThat(row.getInt("ttl")).isLessThanOrEqualTo(123456789);
   }
 
-  @Test
-  void unload_load_preserving_ttl_and_timestamp() throws IOException {
+  @ParameterizedTest
+  @MethodSource
+  void unload_load_preserving_ttl_and_timestamp(
+      DataType cqlType, Object value1, Object value2, String csv1, String csv2) throws IOException {
+
+    checkCqlTypeSupported(cqlType);
 
     session.execute("DROP TABLE IF EXISTS preserve_ttl_timestamp");
     session.execute(
-        "CREATE TABLE preserve_ttl_timestamp (pk1 int, pk2 int, cc1 int, cc2 int, v1 text, v2 text, PRIMARY KEY ((pk1, pk2), cc1, cc2))");
+        String.format(
+            "CREATE TABLE preserve_ttl_timestamp (pk1 int, pk2 int, cc1 int, cc2 int, v1 %s, v2 %s, PRIMARY KEY ((pk1, pk2), cc1, cc2))",
+            cqlType.asCql(true, true), cqlType.asCql(true, true)));
+    TypeCodec<Object> codec = CodecRegistry.DEFAULT.codecFor(cqlType);
     session.execute(
-        "BEGIN BATCH "
-            + "INSERT INTO preserve_ttl_timestamp (pk1, pk2, cc1, cc2, v1) "
-            + "VALUES (1, 2, 3, 4, 'foo') "
-            + "USING TIMESTAMP 1111 AND TTL 111111; "
-            + "INSERT INTO preserve_ttl_timestamp (pk1, pk2, cc1, cc2, v2) "
-            + "VALUES (1, 2, 3, 4, 'bar') "
-            + "USING TIMESTAMP 2222 AND TTL 222222; "
-            + "APPLY BATCH");
+        String.format(
+            "BEGIN BATCH "
+                + "INSERT INTO preserve_ttl_timestamp (pk1, pk2, cc1, cc2, v1) "
+                + "VALUES (1, 2, 3, 4, %s) "
+                + "USING TIMESTAMP 1111 AND TTL 111111; "
+                + "INSERT INTO preserve_ttl_timestamp (pk1, pk2, cc1, cc2, v2) "
+                + "VALUES (1, 2, 3, 4, %s) "
+                + "USING TIMESTAMP 2222 AND TTL 222222; "
+                + "APPLY BATCH",
+            codec.format(value1), codec.format(value2)));
 
     List<String> args =
         Lists.newArrayList(
@@ -1993,9 +2019,16 @@ class CSVConnectorEndToEndCCMIT extends EndToEndCCMITBase {
     Stream<String> line = FileUtils.readAllLinesInDirectoryAsStreamExcludingHeaders(unloadDir);
     assertThat(line)
         .singleElement(InstanceOfAssertFactories.STRING)
-        .containsPattern(
-            "1,2,3,4,foo,1970-01-01T00:00:00\\.001111Z,111\\d\\d\\d,bar,1970-01-01T00:00:00\\.002222Z,222\\d\\d\\d");
+        .contains("1,2,3,4,", csv1, csv2)
+        .containsPattern(",1970-01-01T00:00:00\\.001111Z,111\\d\\d\\d,")
+        .containsPattern(",1970-01-01T00:00:00\\.002222Z,222\\d\\d\\d");
+
+    assertThat(logs)
+        .doesNotHaveMessageContaining("Skipping timestamp preservation")
+        .doesNotHaveMessageContaining("Skipping TTL preservation");
+
     FileUtils.deleteDirectory(logDir);
+    logs.clear();
     session.execute("TRUNCATE preserve_ttl_timestamp");
 
     args =
@@ -2032,12 +2065,237 @@ class CSVConnectorEndToEndCCMIT extends EndToEndCCMITBase {
     assertThat(row.getInt("pk2")).isEqualTo(2);
     assertThat(row.getInt("cc1")).isEqualTo(3);
     assertThat(row.getInt("cc2")).isEqualTo(4);
-    assertThat(row.getString("v1")).isEqualTo("foo");
-    assertThat(row.getString("v2")).isEqualTo("bar");
+    assertThat(row.getObject("v1")).isEqualTo(value1);
+    assertThat(row.getObject("v2")).isEqualTo(value2);
     assertThat(row.getLong("v1w")).isEqualTo(1111L);
     assertThat(row.getLong("v2w")).isEqualTo(2222L);
     assertThat(row.getInt("v1t")).isLessThanOrEqualTo(111111).isGreaterThan(111000);
     assertThat(row.getInt("v2t")).isLessThanOrEqualTo(222222).isGreaterThan(222000);
+
+    assertThat(logs)
+        .doesNotHaveMessageContaining("Skipping timestamp preservation")
+        .doesNotHaveMessageContaining("Skipping TTL preservation");
+  }
+
+  @SuppressWarnings("unused")
+  Stream<Arguments> unload_load_preserving_ttl_and_timestamp() throws UnknownHostException {
+    session.execute("CREATE TYPE IF NOT EXISTS preserve_ttl_timestamp_udt (f1 int, f2 text)");
+    TupleType tupleType = DataTypes.tupleOf(DataTypes.INT, DataTypes.TEXT);
+    UserDefinedType udt =
+        session
+            .getKeyspace()
+            .flatMap(ks -> session.getMetadata().getKeyspace(ks))
+            .flatMap(ks -> ks.getUserDefinedType("preserve_ttl_timestamp_udt"))
+            .orElseThrow(IllegalStateException::new);
+    return Stream.of(
+        Arguments.of(DataTypes.ASCII, "abc", "def", "abc", "def"),
+        Arguments.of(DataTypes.BIGINT, 123L, 456L, "123", "456"),
+        Arguments.of(
+            DataTypes.BLOB,
+            ByteBuffer.wrap(new byte[] {1, 2}),
+            ByteBuffer.wrap(new byte[] {3, 4}),
+            "AQI=",
+            "AwQ="),
+        Arguments.of(DataTypes.BOOLEAN, true, false, "1", "0"),
+        //    Arguments.of(DataTypes.COUNTER, v1, v2, csv1, csv2),
+        Arguments.of(
+            DataTypes.DECIMAL, new BigDecimal("12.34"), new BigDecimal("56.78"), "12.34", "56.78"),
+        Arguments.of(DataTypes.DOUBLE, 12.34, 56.78, "12.34", "56.78"),
+        Arguments.of(DataTypes.FLOAT, 12.34f, 56.78f, "12.34", "56.78"),
+        Arguments.of(DataTypes.INT, 12, 34, "12", "34"),
+        Arguments.of(
+            DataTypes.TIMESTAMP,
+            Instant.parse("2021-01-11T11:01:01Z"),
+            Instant.parse("2022-02-22T22:02:02Z"),
+            "2021-01-11T11:01:01Z",
+            "2022-02-22T22:02:02Z"),
+        Arguments.of(
+            DataTypes.UUID,
+            Uuids.startOf(123456789),
+            Uuids.endOf(123456789),
+            Uuids.startOf(123456789).toString(),
+            Uuids.endOf(123456789).toString()),
+        Arguments.of(
+            DataTypes.VARINT, new BigInteger("1234"), new BigInteger("5678"), "1234", "5678"),
+        Arguments.of(
+            DataTypes.TIMEUUID,
+            Uuids.startOf(123456789),
+            Uuids.endOf(123456789),
+            Uuids.startOf(123456789).toString(),
+            Uuids.endOf(123456789).toString()),
+        Arguments.of(
+            DataTypes.INET,
+            InetAddress.getByAddress(new byte[] {1, 2, 3, 4}),
+            InetAddress.getByAddress(new byte[] {5, 6, 7, 8}),
+            "1.2.3.4",
+            "5.6.7.8"),
+        Arguments.of(
+            DataTypes.DATE,
+            LocalDate.of(2021, 1, 11),
+            LocalDate.of(2022, 2, 22),
+            "2021-01-11",
+            "2022-02-22"),
+        Arguments.of(DataTypes.TEXT, "abc", "def", "abc", "def"),
+        Arguments.of(
+            DataTypes.TIME, LocalTime.of(1, 2, 3), LocalTime.of(4, 5, 6), "01:02:03", "04:05:06"),
+        Arguments.of(DataTypes.SMALLINT, (short) 12, (short) 34, "12", "34"),
+        Arguments.of(DataTypes.TINYINT, (byte) 12, (byte) 34, "12", "34"),
+        Arguments.of(
+            DataTypes.DURATION,
+            CqlDuration.newInstance(1, 2, 34),
+            CqlDuration.newInstance(5, 6, 78),
+            "1mo2d34ns",
+            "5mo6d78ns"),
+        Arguments.of(
+            tupleType,
+            tupleType.newValue(12, "ab"),
+            tupleType.newValue(34, "cd"),
+            "[12,\\\"ab\\\"]",
+            "[34,\\\"cd\\\"]"),
+        Arguments.of(
+            udt.copy(true), // only frozen UDTs work
+            udt.newValue(12, "ab"),
+            udt.newValue(34, "cd"),
+            "{\\\"f1\\\":12,\\\"f2\\\":\\\"ab\\\"}",
+            "{\\\"f1\\\":34,\\\"f2\\\":\\\"cd\\\"}"));
+  }
+
+  @ParameterizedTest
+  @MethodSource
+  void unload_load_preserving_ttl_and_timestamp_unsupported_types(DataType cqlType, Object value)
+      throws IOException {
+
+    checkCqlTypeSupported(cqlType);
+
+    session.execute("DROP TABLE IF EXISTS preserve_ttl_timestamp");
+    String typeAsCql = cqlType.asCql(true, true);
+    session.execute(
+        String.format(
+            "CREATE TABLE preserve_ttl_timestamp (pk1 int, pk2 int, cc1 int, cc2 int, v1 %s, PRIMARY KEY ((pk1, pk2), cc1, cc2))",
+            typeAsCql));
+    TypeCodec<Object> codec = CodecRegistry.DEFAULT.codecFor(cqlType);
+    session.execute(
+        String.format(
+            "INSERT INTO preserve_ttl_timestamp (pk1, pk2, cc1, cc2, v1) "
+                + "VALUES (1, 2, 3, 4, %s) "
+                + "USING TIMESTAMP 1111 AND TTL 111111",
+            codec.format(value)));
+
+    List<String> args =
+        Lists.newArrayList(
+            "unload",
+            "--log.directory",
+            quoteJson(logDir),
+            "--connector.csv.url",
+            quoteJson(unloadDir),
+            "--connector.csv.header",
+            "true",
+            "--schema.keyspace",
+            session.getKeyspace().get().asInternal(),
+            "--schema.table",
+            "preserve_ttl_timestamp",
+            "-timestamp",
+            "true",
+            "-ttl",
+            "true");
+
+    ExitStatus status = new DataStaxBulkLoader(addCommonSettings(args)).run();
+    assertStatus(status, STATUS_OK);
+
+    Stream<String> line = FileUtils.readAllLinesInDirectoryAsStreamExcludingHeaders(unloadDir);
+    assertThat(line).singleElement(InstanceOfAssertFactories.STRING).doesNotContain("1111", "2222");
+
+    assertThat(logs)
+        .hasMessageContaining(
+            String.format(
+                "Skipping timestamp preservation for column v1: this feature is not supported for CQL type %s",
+                typeAsCql))
+        .hasMessageContaining(
+            String.format(
+                "Skipping TTL preservation for column v1: this feature is not supported for CQL type %s",
+                typeAsCql));
+
+    FileUtils.deleteDirectory(logDir);
+    logs.clear();
+    session.execute("TRUNCATE preserve_ttl_timestamp");
+
+    args =
+        Lists.newArrayList(
+            "load",
+            "--log.directory",
+            quoteJson(logDir),
+            "--connector.csv.url",
+            quoteJson(unloadDir),
+            "--connector.csv.header",
+            "true",
+            "--schema.keyspace",
+            session.getKeyspace().get().asInternal(),
+            "--schema.table",
+            "preserve_ttl_timestamp",
+            "-timestamp",
+            "true",
+            "-ttl",
+            "true");
+
+    status = new DataStaxBulkLoader(addCommonSettings(args)).run();
+    assertStatus(status, STATUS_OK);
+
+    ResultSet rs =
+        session.execute(
+            "SELECT pk1, pk2, cc1, cc2, v1 "
+                + "FROM preserve_ttl_timestamp "
+                + "WHERE pk1 = 1 AND pk2 = 2 AND cc1 = 3 AND cc2 = 4");
+    Row row = rs.one();
+    assertThat(row).isNotNull();
+    assertThat(row.getInt("pk1")).isEqualTo(1);
+    assertThat(row.getInt("pk2")).isEqualTo(2);
+    assertThat(row.getInt("cc1")).isEqualTo(3);
+    assertThat(row.getInt("cc2")).isEqualTo(4);
+    assertThat(row.getObject("v1")).isEqualTo(value);
+
+    assertThat(logs)
+        .hasMessageContaining(
+            String.format(
+                "Skipping timestamp preservation for column v1: this feature is not supported for CQL type %s",
+                typeAsCql))
+        .hasMessageContaining(
+            String.format(
+                "Skipping TTL preservation for column v1: this feature is not supported for CQL type %s",
+                typeAsCql));
+  }
+
+  @SuppressWarnings("unused")
+  Stream<Arguments> unload_load_preserving_ttl_and_timestamp_unsupported_types() {
+    session.execute("CREATE TYPE IF NOT EXISTS preserve_ttl_timestamp_udt (f1 int, f2 text)");
+    UserDefinedType udt =
+        session
+            .getKeyspace()
+            .flatMap(ks -> session.getMetadata().getKeyspace(ks))
+            .flatMap(ks -> ks.getUserDefinedType("preserve_ttl_timestamp_udt"))
+            .orElseThrow(IllegalStateException::new);
+    return Stream.of(
+        Arguments.of(DataTypes.listOf(DataTypes.INT), ImmutableList.of(12, 34)),
+        Arguments.of(DataTypes.setOf(DataTypes.INT), ImmutableSet.of(12, 34)),
+        Arguments.of(
+            DataTypes.mapOf(DataTypes.INT, DataTypes.TEXT), ImmutableMap.of(12, "ab", 34, "cd")),
+        Arguments.of(DataTypes.frozenListOf(DataTypes.INT), ImmutableList.of(12, 34)),
+        Arguments.of(DataTypes.frozenSetOf(DataTypes.INT), ImmutableSet.of(12, 34)),
+        Arguments.of(
+            DataTypes.frozenMapOf(DataTypes.INT, DataTypes.TEXT),
+            ImmutableMap.of(12, "ab", 34, "cd")),
+        Arguments.of(udt, udt.newValue(12, "ab")) // non-frozen UDT
+        );
+  }
+
+  private void checkCqlTypeSupported(DataType cqlType) {
+    Assumptions.assumeTrue(
+        CQLUtils.isCqlTypeSupported(
+            cqlType, session.getContext().getProtocolVersion(), ccm.getCassandraVersion()),
+        String.format(
+            "CQL type %s not compatible with C* %s and protocol version %s",
+            cqlType.asCql(true, true),
+            ccm.getCassandraVersion(),
+            session.getContext().getProtocolVersion()));
   }
 
   @Test

--- a/tests/src/main/java/com/datastax/oss/dsbulk/tests/RemoteClusterExtension.java
+++ b/tests/src/main/java/com/datastax/oss/dsbulk/tests/RemoteClusterExtension.java
@@ -17,6 +17,7 @@ package com.datastax.oss.dsbulk.tests;
 
 import com.datastax.oss.driver.api.core.CqlSession;
 import com.datastax.oss.driver.api.core.metadata.EndPoint;
+import com.datastax.oss.driver.api.core.session.Session;
 import com.datastax.oss.dsbulk.tests.driver.factory.SessionFactory;
 import java.lang.reflect.Parameter;
 import java.util.HashSet;
@@ -44,7 +45,8 @@ public abstract class RemoteClusterExtension implements AfterAllCallback, Parame
       ParameterContext parameterContext, ExtensionContext extensionContext)
       throws ParameterResolutionException {
     Class<?> type = parameterContext.getParameter().getType();
-    return type.isAssignableFrom(CqlSession.class);
+    // type must implement Session and be a supertype of CqlSession
+    return Session.class.isAssignableFrom(type) && type.isAssignableFrom(CqlSession.class);
   }
 
   @Override

--- a/tests/src/main/java/com/datastax/oss/dsbulk/tests/ccm/CCMExtension.java
+++ b/tests/src/main/java/com/datastax/oss/dsbulk/tests/ccm/CCMExtension.java
@@ -92,7 +92,7 @@ public class CCMExtension extends RemoteClusterExtension
       ParameterContext parameterContext, ExtensionContext extensionContext)
       throws ParameterResolutionException {
     Class<?> type = parameterContext.getParameter().getType();
-    return type.isAssignableFrom(DefaultCCMCluster.class)
+    return type.equals(CCMCluster.class)
         || super.supportsParameter(parameterContext, extensionContext);
   }
 
@@ -102,7 +102,7 @@ public class CCMExtension extends RemoteClusterExtension
       throws ParameterResolutionException {
     Parameter parameter = parameterContext.getParameter();
     Class<?> type = parameter.getType();
-    if (type.isAssignableFrom(DefaultCCMCluster.class)) {
+    if (type.equals(CCMCluster.class)) {
       CCMCluster ccm = getOrCreateCCM(extensionContext);
       LOGGER.debug(String.format("Returning %s for parameter %s", ccm, parameter));
       return ccm;

--- a/tests/src/main/java/com/datastax/oss/dsbulk/tests/utils/CQLUtils.java
+++ b/tests/src/main/java/com/datastax/oss/dsbulk/tests/utils/CQLUtils.java
@@ -16,7 +16,12 @@
 package com.datastax.oss.dsbulk.tests.utils;
 
 import com.datastax.oss.driver.api.core.CqlIdentifier;
+import com.datastax.oss.driver.api.core.ProtocolVersion;
+import com.datastax.oss.driver.api.core.Version;
 import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.core.type.DataType;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
+import com.datastax.oss.protocol.internal.ProtocolConstants;
 
 public abstract class CQLUtils {
 
@@ -63,5 +68,23 @@ public abstract class CQLUtils {
   public static SimpleStatement truncateTable(CqlIdentifier keyspace, CqlIdentifier table) {
     return SimpleStatement.newInstance(
         "TRUNCATE " + keyspace.asCql(true) + "." + table.asCql(true));
+  }
+
+  public static boolean isCqlTypeSupported(
+      DataType cqlType, ProtocolVersion version, Version cassandraVersion) {
+    switch (cqlType.getProtocolCode()) {
+      case ProtocolConstants.DataType.DATE:
+      case ProtocolConstants.DataType.TIME:
+      case ProtocolConstants.DataType.SMALLINT:
+      case ProtocolConstants.DataType.TINYINT:
+        return version.getCode() >= 4;
+      case ProtocolConstants.DataType.DURATION:
+        return version.getCode() >= 5;
+      case ProtocolConstants.DataType.UDT:
+        // https://issues.apache.org/jira/browse/CASSANDRA-7423
+        return ((UserDefinedType) cqlType).isFrozen()
+            || cassandraVersion.compareTo(Version.parse("3.6")) >= 0;
+    }
+    return true;
   }
 }

--- a/workflow/commons/src/main/java/com/datastax/oss/dsbulk/workflow/commons/settings/SchemaSettings.java
+++ b/workflow/commons/src/main/java/com/datastax/oss/dsbulk/workflow/commons/settings/SchemaSettings.java
@@ -50,6 +50,10 @@ import com.datastax.oss.driver.api.core.metadata.schema.RelationMetadata;
 import com.datastax.oss.driver.api.core.metadata.schema.TableMetadata;
 import com.datastax.oss.driver.api.core.metadata.schema.ViewMetadata;
 import com.datastax.oss.driver.api.core.type.DataTypes;
+import com.datastax.oss.driver.api.core.type.ListType;
+import com.datastax.oss.driver.api.core.type.MapType;
+import com.datastax.oss.driver.api.core.type.SetType;
+import com.datastax.oss.driver.api.core.type.UserDefinedType;
 import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
 import com.datastax.oss.driver.shaded.guava.common.base.Preconditions;
 import com.datastax.oss.driver.shaded.guava.common.base.Predicates;
@@ -643,13 +647,11 @@ public class SchemaSettings {
                   column -> {
                     CQLWord colName = CQLWord.fromCqlIdentifier(column.getName());
                     List<CQLFragment> cols = Lists.newArrayList(colName);
-                    if (!isCounterTable()
-                        && schemaGenerationStrategy.isMapping()
-                        && !table.getPrimaryKey().contains(column)) {
-                      if (preserveTimestamp) {
+                    if (schemaGenerationStrategy.isMapping()) {
+                      if (preserveTimestamp && checkWritetimeTtlSupported(column, WRITETIME)) {
                         cols.add(new FunctionCall(null, WRITETIME, colName));
                       }
-                      if (preserveTtl) {
+                      if (preserveTtl && checkWritetimeTtlSupported(column, TTL)) {
                         cols.add(new FunctionCall(null, TTL, colName));
                       }
                     }
@@ -773,6 +775,37 @@ public class SchemaSettings {
       builder.add(session.prepare(childStatement));
     }
     return builder.build();
+  }
+
+  private boolean checkWritetimeTtlSupported(ColumnMetadata col, CQLWord functionName) {
+    if (isCounterTable() || table.getPrimaryKey().contains(col)) {
+      // Counter tables and primary key columns don't have timestamps and TTLs.
+      return false;
+    }
+    boolean supported;
+    if (col.getType() instanceof ListType
+        || col.getType() instanceof SetType
+        || col.getType() instanceof MapType) {
+      // Collection support for WRITETIME and TTL varies depending on the Cassandra and DSE versions
+      // in use, but in all cases, even if the export is possible, it wouldn't be possible to import
+      // the data back to Cassandra. This is valid for frozen collections as well.
+      supported = false;
+    } else if (col.getType() instanceof UserDefinedType) {
+      // While most C* versions (all?) accept WRITETIME and TTL functions on UDTs, the functions
+      // (always?) return NULL for non-frozen ones. Therefore, we only consider frozen UDTs as
+      // supported. Also note that all tuples are accepted, since they are frozen by nature.
+      supported = ((UserDefinedType) col.getType()).isFrozen();
+    } else {
+      supported = true;
+    }
+    if (!supported) {
+      LOGGER.warn(
+          "Skipping {} preservation for column {}: this feature is not supported for CQL type {}.",
+          functionName == WRITETIME ? "timestamp" : "TTL",
+          col.getName().asCql(true),
+          col.getType().asCql(true, true));
+    }
+    return supported;
   }
 
   private boolean isDSESearchPseudoColumn(ColumnMetadata col) {

--- a/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/settings/SchemaSettingsTest.java
+++ b/workflow/commons/src/test/java/com/datastax/oss/dsbulk/workflow/commons/settings/SchemaSettingsTest.java
@@ -2555,8 +2555,44 @@ class SchemaSettingsTest {
         .hasMessageContaining("ttl(*) is not allowed when schema.preserveTtl is true");
   }
 
-  // Function calls are not allowed when updating a counter table.
-  // Cannot set TTL or timestamp when updating a counter table.
+  // TODO Function calls are not allowed when updating a counter table.
+  // TODO Cannot set TTL or timestamp when updating a counter table.
+
+  @Test
+  void should_not_preserve_timestamp_when_unsupported_type() {
+    when(col3.getType()).thenReturn(DataTypes.listOf(DataTypes.TEXT));
+    Config config =
+        TestConfigUtils.createTestConfig(
+            "dsbulk.schema", "keyspace", "ks", "table", "t1", "preserveTimestamp", true);
+    SchemaSettings settings = new SchemaSettings(config, READ_AND_MAP);
+    settings.init(session, true, true);
+    settings.createReadResultMapper(session, recordMetadata, codecFactory, true);
+    assertThat(getInternalState(settings, "query"))
+        .isEqualTo(
+            "SELECT c1, \"COL 2\", writetime(\"COL 2\") AS \"writetime(COL 2)\", c3 "
+                + "FROM ks.t1 WHERE token(c1) > :start AND token(c1) <= :end");
+    assertThat(logs)
+        .hasMessageContaining(
+            "Skipping timestamp preservation for column c3: this feature is not supported for CQL type list<text>");
+  }
+
+  @Test
+  void should_not_preserve_ttl_when_unsupported_type() {
+    when(col3.getType()).thenReturn(DataTypes.listOf(DataTypes.TEXT));
+    Config config =
+        TestConfigUtils.createTestConfig(
+            "dsbulk.schema", "keyspace", "ks", "table", "t1", "preserveTtl", true);
+    SchemaSettings settings = new SchemaSettings(config, READ_AND_MAP);
+    settings.init(session, true, true);
+    settings.createReadResultMapper(session, recordMetadata, codecFactory, true);
+    assertThat(getInternalState(settings, "query"))
+        .isEqualTo(
+            "SELECT c1, \"COL 2\", ttl(\"COL 2\") AS \"ttl(COL 2)\", c3 "
+                + "FROM ks.t1 WHERE token(c1) > :start AND token(c1) <= :end");
+    assertThat(logs)
+        .hasMessageContaining(
+            "Skipping TTL preservation for column c3: this feature is not supported for CQL type list<text>");
+  }
 
   @ParameterizedTest
   @MethodSource

